### PR TITLE
Better read from cache buffer

### PR DESCRIPTION
**Pull request by kssenii**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Better read from cache.

### CODE REVIEW

File: src/Disks/IO/CachedReadBufferFromRemoteFS.cpp [hunk 1]
```diff
@@ -362,7 +386,7 @@ SeekableReadBufferPtr CachedReadBufferFromRemoteFS::getImplementationBuffer(File
         case ReadType::CACHED:
         {
 #ifndef NDEBUG
-            auto * file_reader = assert_cast<ReadBufferFromFile *>(read_buffer_for_file_segment.get());
+            auto * file_reader = dynamic_cast<ReadBufferFromFileDescriptor *>(read_buffer_for_file_segment.get());
```
Comment (kssenii) [inline src/Disks/IO/CachedReadBufferFromRemoteFS.cpp:389]: This.

Now 99% of whole usage of this `CachedBufferFromRemoteFS` is consumed by the `next()` call of the underlying buffer from local fs. See https://pastila.nl/?009d10d1/7098f333e2cc02de544ad7d3409216e9.

File: src/Disks/IO/CachedReadBufferFromRemoteFS.cpp [hunk 2]
```diff
@@ -109,9 +125,16 @@ void CachedReadBufferFromRemoteFS::initialize(size_t offset, size_t size)
...
 {
     auto path = cache->getPathInLocalCache(cache_key, offset);
-    auto buf = std::make_shared<ReadBufferFromFile>(path, settings.local_fs_buffer_size);
-    if (buf->size() == 0)
+
+    ReadSettings local_read_settings{settings};
+    /// Do not allow to use asynchronous version of LocalFSReadMethod.
+    local_read_settings.local_fs_method = LocalFSReadMethod::pread;
+
+    auto buf = createReadBufferFromFileBase(path, local_read_settings);
```
[the remark this answers is not in the document]
Comment (kssenii) [reply src/Disks/IO/CachedReadBufferFromRemoteFS.cpp:133]: yes

File: src/Common/ProfileEvents.cpp [hunk 3]
```diff
@@ -260,10 +260,12 @@
...
-    M(RemoteFSReadMicroseconds, "Time of reading from remote filesystem.") \
-    M(RemoteFSReadBytes, "Read bytes from remote filesystem.") \
-    M(RemoteFSCacheReadBytes, "Read bytes from cache of remote filesystem.") \
-    M(RemoteFSCacheDownloadBytes, "Bytes downloaded to cache from remote filesystem.") \
+    M(CachedReadBufferReadFromSourceMicroseconds, "Time reading from source (remote fs, etc)") \
+    M(CachedReadBufferReadFromCacheMicroseconds, "Time reading from cache") \
+    M(CachedReadBufferReadFromSourceBytes, "Bytes read from source (remote fs, etc)") \
+    M(CachedReadBufferReadFromCacheBytes, "Bytes read from cache") \
+    M(CachedReadBufferCacheWriteBytes, "Bytes written from source to cache") \
+    M(CachedReadBufferCacheWriteMicroseconds, "Time spent writing data into cache") \
```
Comment (tavplubix) [inline src/Common/ProfileEvents.cpp:268]: These descriptions are not user-readable (what cache? what source?)

**Review by tavplubix — APPROVED** [review]
Everything else LGTM

**Comment by kssenii** [comment]
> Stateless tests (thread, actions) [3/3] — Tests are not finished, fail: 2

https://github.com/ClickHouse/ClickHouse/issues/37184

**Pull request merged**